In [1]:
from warnings import filterwarnings
filterwarnings('ignore')

### Step 1: Read the data set

In [2]:
import pandas as pd

df = pd.read_csv('training_set.csv')
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

### step 2 : basic data quality checks

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [5]:
m = df.isna().sum()
m[m>0]

LotFrontage      259
Alley           1369
MasVnrType       872
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

### Step 3: Seperate X and Y()

In [8]:
X= df.drop(columns=['Id', 'SalePrice'])
Y = df[['SalePrice']]

In [9]:
X.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... ScreenPorch PoolArea PoolQC Fence  \
0         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
1         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   
2         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
3         Lvl    AllPub    Corner  ...           0        0    NaN   NaN   
4         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0       2    2008        WD         Normal  
1         NaN       0       5    2007        WD         Normal  
2         NaN       0       9    2008        WD         Normal  
3         NaN       0       2    2006        WD        Abnorml  
4         NaN       0      12    2008        WD         Normal  

[5 rows x 79 columns]

In [10]:
Y.head()

SalePrice
0     208500
1     181500
2     223500
3     140000
4     250000

### Step 4 : Preprocessing on X

In [11]:
cat = list(X.columns[X.dtypes=='object'])
con = list(X.columns[X.dtypes!='object'])

In [12]:
cat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [13]:
con

['MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold']

In [14]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OrdinalEncoder
from sklearn.compose import ColumnTransformer


In [15]:
num_pipe1 = Pipeline(steps=[('impute', SimpleImputer(strategy='mean')), ('scaler', StandardScaler())])

In [16]:
cat_pipe1 = Pipeline(steps=[('impute', SimpleImputer(strategy='constant', fill_value='Not_Avail')), ('ordinal', OrdinalEncoder())])

In [17]:
pre1 = ColumnTransformer([('num', num_pipe1, con), ('cat', cat_pipe1, cat)]).set_output(transform='pandas')

In [18]:
X_pre1 = pre1.fit_transform(X)
X_pre1

num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
0            0.073375         -0.229372     -0.207142          0.651479   
1           -0.872563          0.451936     -0.091886         -0.071836   
2            0.073375         -0.093110      0.073480          0.651479   
3            0.309859         -0.456474     -0.096897          0.651479   
4            0.073375          0.633618      0.375148          1.374795   
...               ...               ...           ...               ...   
1455         0.073375         -0.365633     -0.260560         -0.071836   
1456        -0.872563          0.679039      0.266407         -0.071836   
1457         0.309859         -0.183951     -0.147810          0.651479   
1458        -0.872563         -0.093110     -0.080160         -0.795151   
1459        -0.872563          0.224833     -0.058112         -0.795151   

      num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
0            -0.517200        1.050994           0.878668         0.511418   
1             2.179628        0.156734          -0.429577        -0.574410   
2            -0.517200        0.984752           0.830215         0.323060   
3            -0.517200       -1.863632          -0.720298        -0.574410   
4            -0.517200        0.951632           0.733308         1.364570   
...                ...             ...                ...              ...   
1455         -0.517200        0.918511           0.733308        -0.574410   
1456          0.381743        0.222975           0.151865         0.084843   
1457          3.078570       -1.002492           1.024029        -0.574410   
1458          0.381743       -0.704406           0.539493        -0.574410   
1459          0.381743       -0.207594          -0.962566        -0.574410   

      num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__GarageType  \
0            0.575425        -0.288653  ...              1.0   
1            1.171992        -0.288653  ...              1.0   
2            0.092907        -0.288653  ...              1.0   
3           -0.499274        -0.288653  ...              5.0   
4            0.463568        -0.288653  ...              1.0   
...               ...              ...  ...              ...   
1455        -0.973018        -0.288653  ...              1.0   
1456         0.759659         0.722112  ...              1.0   
1457        -0.369871        -0.288653  ...              1.0   
1458        -0.865548         6.092188  ...              1.0   
1459         0.847389         1.509640  ...              1.0   

      cat__GarageFinish  cat__GarageQual  cat__GarageCond  cat__PavedDrive  \
0                   2.0              5.0              5.0              2.0   
1                   2.0              5.0              5.0              2.0   
2                   2.0              5.0              5.0              2.0   
3                   3.0              5.0              5.0              2.0   
4                   2.0              5.0              5.0              2.0   
...                 ...              ...              ...              ...   
1455                2.0              5.0              5.0              2.0   
1456                3.0              5.0              5.0              2.0   
1457                2.0              5.0              5.0              2.0   
1458                3.0              5.0              5.0              2.0   
1459                0.0              5.0              5.0              2.0   

      cat__PoolQC  cat__Fence  cat__MiscFeature  cat__SaleType  \
0             3.0         4.0               1.0            8.0   
1             3.0         4.0               1.0            8.0   
2             3.0         4.0               1.0            8.0   
3             3.0         4.0               1.0            8.0   
4             3.0         4.0               1.0            8.0   
...           ...         ...               ...            ...   


In [19]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SequentialFeatureSelector


In [20]:
lr = LinearRegression()
sel = SequentialFeatureSelector(estimator=lr, n_features_to_select='auto', direction='backward')
sel.fit(X_pre1, Y)
sel_cols = sel.get_feature_names_out()
sel_cols

array(['num__MSSubClass', 'num__LotArea', 'num__OverallQual',
       'num__OverallCond', 'num__YearBuilt', 'num__MasVnrArea',
       'num__BsmtFinSF1', 'num__BsmtFinSF2', 'num__BsmtUnfSF',
       'num__TotalBsmtSF', 'num__GrLivArea', 'num__BsmtFullBath',
       'num__TotRmsAbvGrd', 'num__Fireplaces', 'num__GarageCars',
       'num__WoodDeckSF', 'num__EnclosedPorch', 'num__ScreenPorch',
       'num__PoolArea', 'num__YrSold', 'cat__LandContour',
       'cat__Neighborhood', 'cat__BldgType', 'cat__HouseStyle',
       'cat__RoofMatl', 'cat__Exterior1st', 'cat__MasVnrType',
       'cat__ExterQual', 'cat__BsmtQual', 'cat__BsmtCond',
       'cat__BsmtExposure', 'cat__HeatingQC', 'cat__KitchenQual',
       'cat__Functional', 'cat__FireplaceQu', 'cat__GarageFinish',
       'cat__GarageQual', 'cat__PavedDrive', 'cat__MiscFeature',
       'cat__SaleCondition'], dtype=object)

In [21]:
len(sel_cols)

40

In [22]:
imp_cols = []
for i in sel_cols:
    s= i.split('__')[1]
    imp_cols.append(s)

In [23]:
X_sel = X[imp_cols]
X_sel

MSSubClass  LotArea  OverallQual  OverallCond  YearBuilt  MasVnrArea  \
0             60     8450            7            5       2003       196.0   
1             20     9600            6            8       1976         0.0   
2             60    11250            7            5       2001       162.0   
3             70     9550            7            5       1915         0.0   
4             60    14260            8            5       2000       350.0   
...          ...      ...          ...          ...        ...         ...   
1455          60     7917            6            5       1999         0.0   
1456          20    13175            6            6       1978       119.0   
1457          70     9042            7            9       1941         0.0   
1458          20     9717            5            6       1950         0.0   
1459          20     9937            5            6       1965         0.0   

      BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  ...  BsmtExposure  \
0            706           0        150          856  ...            No   
1            978           0        284         1262  ...            Gd   
2            486           0        434          920  ...            Mn   
3            216           0        540          756  ...            No   
4            655           0        490         1145  ...            Av   
...          ...         ...        ...          ...  ...           ...   
1455           0           0        953          953  ...            No   
1456         790         163        589         1542  ...            No   
1457         275           0        877         1152  ...            No   
1458          49        1029          0         1078  ...            Mn   
1459         830         290        136         1256  ...            No   

      HeatingQC  KitchenQual  Functional  FireplaceQu  GarageFinish  \
0            Ex           Gd         Typ          NaN           RFn   
1            Ex           TA         Typ           TA           RFn   
2            Ex           Gd         Typ           TA           RFn   
3            Gd           Gd         Typ           Gd           Unf   
4            Ex           Gd         Typ           TA           RFn   
...         ...          ...         ...          ...           ...   
1455         Ex           TA         Typ           TA           RFn   
1456         TA           TA        Min1           TA           Unf   
1457         Ex           Gd         Typ           Gd           RFn   
1458         Gd           Gd         Typ          NaN           Unf   
1459         Gd           TA         Typ          NaN           Fin   

      GarageQual  PavedDrive  MiscFeature  SaleCondition  
0             TA           Y          NaN         Normal  
1             TA           Y          NaN         Normal  
2             TA           Y          NaN         Normal  
3             TA           Y          NaN        Abnorml  
4             TA           Y          NaN         Normal  
...          ...         ...          ...            ...  
1455          TA           Y          NaN         Normal  
1456          TA           Y          NaN         Normal  
1457          TA           Y         Shed         Normal  
1458          TA           Y          NaN         Normal  
1459          TA           Y          NaN         Normal  

[1460 rows x 40 columns]

In [26]:
cat_sel = list(X_sel.columns[X_sel.dtypes=='object'])
con_sel = list(X_sel.columns[X_sel.dtypes!='object'])

In [28]:
from sklearn.preprocessing import OneHotEncoder

In [27]:
num_pipe2 = Pipeline(steps=[('impute', SimpleImputer(strategy='mean')), ('scaler', StandardScaler())])

In [29]:
cat_pipe2 = Pipeline(steps=[('impute', SimpleImputer(strategy='constant', fill_value='Not_avail')), ('ohe', OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [32]:
pre2 = ColumnTransformer([('num', num_pipe2, con_sel), ('cat', cat_pipe2, cat_sel)]).set_output(transform='pandas')

In [33]:
X_sel_pre = pre2.fit_transform(X_sel)
X_sel_pre

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
0            0.073375     -0.207142          0.651479         -0.517200   
1           -0.872563     -0.091886         -0.071836          2.179628   
2            0.073375      0.073480          0.651479         -0.517200   
3            0.309859     -0.096897          0.651479         -0.517200   
4            0.073375      0.375148          1.374795         -0.517200   
...               ...           ...               ...               ...   
1455         0.073375     -0.260560         -0.071836         -0.517200   
1456        -0.872563      0.266407         -0.071836          0.381743   
1457         0.309859     -0.147810          0.651479          3.078570   
1458        -0.872563     -0.080160         -0.795151          0.381743   
1459        -0.872563     -0.058112         -0.795151          0.381743   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
0           1.050994         0.511418         0.575425        -0.288653   
1           0.156734        -0.574410         1.171992        -0.288653   
2           0.984752         0.323060         0.092907        -0.288653   
3          -1.863632        -0.574410        -0.499274        -0.288653   
4           0.951632         1.364570         0.463568        -0.288653   
...              ...              ...              ...              ...   
1455        0.918511        -0.574410        -0.973018        -0.288653   
1456        0.222975         0.084843         0.759659         0.722112   
1457       -1.002492        -0.574410        -0.369871        -0.288653   
1458       -0.704406        -0.574410        -0.865548         6.092188   
1459       -0.207594        -0.574410         0.847389         1.509640   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__MiscFeature_Not_avail  \
0          -0.944591         -0.459303  ...                         1.0   
1          -0.641228          0.466465  ...                         1.0   
2          -0.301643         -0.313369  ...                         1.0   
3          -0.061670         -0.687324  ...                         1.0   
4          -0.174865          0.199680  ...                         1.0   
...              ...               ...  ...                         ...   
1455        0.873321         -0.238122  ...                         1.0   
1456        0.049262          1.104925  ...                         1.0   
1457        0.701265          0.215641  ...                         0.0   
1458       -1.284176          0.046905  ...                         1.0   
1459       -0.976285          0.452784  ...                         1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  \
0                       0.0                    0.0                    0.0   
1                       0.0                    0.0                    0.0   
2                       0.0                    0.0                    0.0   
3                       0.0                    0.0                    0.0   
4                       0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
1455                    0.0                    0.0                    0.0   
1456                    0.0                    0.0                    0.0   
1457                    0.0                    1.0                    0.0   
1458                    0.0                    0.0                    0.0   
1459                    0.0                    0.0                    0.0   

      cat__SaleCondition_Abnorml  cat__SaleCondition_AdjLand  \
0                            0.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            1.0                         0.0   
4                            0.0                         0

### Step 5: Perform train and test split

In [35]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(X_sel_pre, Y, test_size=0.2, random_state=21)

In [36]:
xtrain.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
710         -0.636078     -0.640101         -2.241782          0.381743   
1098        -0.163109     -0.452686         -1.518467          0.381743   
1286        -0.872563     -0.072844         -0.071836         -0.517200   
992          0.073375     -0.075851         -0.071836          2.179628   
631          1.492282     -0.593999          1.374795         -0.517200   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
710        -1.201217        -0.574410        -0.973018        -0.288653   
1098       -1.168096        -0.574410         0.500854        -0.288653   
1286       -0.273836         1.924104         0.274948         0.213629   
992        -0.240715         1.309170         0.202570         0.436865   
631         1.150356         0.023903        -0.920380        -0.288653   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__MiscFeature_Not_avail  \
710        -0.672923         -1.795509  ...                         1.0   
1098       -1.284176         -0.878862  ...                         1.0   
1286        0.250749          0.616959  ...                         1.0   
992        -0.901577         -0.536830  ...                         1.0   
631         2.179592          1.132288  ...                         1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  \
710                     0.0                    0.0                    0.0   
1098                    0.0                    0.0                    0.0   
1286                    0.0                    0.0                    0.0   
992                     0.0                    0.0                    0.0   
631                     0.0                    0.0                    0.0   

      cat__SaleCondition_Abnorml  cat__SaleCondition_AdjLand  \
710                          0.0                         0.0   
1098                         0.0                         0.0   
1286                         0.0                         0.0   
992                          0.0                         0.0   
631                          0.0                         0.0   

      cat__SaleCondition_Alloca  cat__SaleCondition_Family  \
710                         0.0                        0.0   
1098                        0.0                        0.0   
1286                        0.0                        0.0   
992                         0.0                        0.0   
631                         0.0                        0.0   

      cat__SaleCondition_Normal  cat__SaleCondition_Partial  
710                         1.0                         0.0  
1098                        1.0                         0.0  
1286                        1.0                         0.0  
992                         1.0                         0.0  
631                         1.0                         0.0  

[5 rows x 154 columns]

In [37]:
xtest.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
880         -0.872563     -0.350058         -0.795151         -0.517200   
605          0.073375      0.309002          0.651479          0.381743   
1166        -0.872563     -0.004192          1.374795         -0.517200   
216         -0.872563     -0.207142          0.651479         -0.517200   
970         -0.163109      0.028380         -1.518467         -1.416142   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
880         1.117235        -0.574410         1.176379        -0.288653   
605        -0.207594         0.400620         0.022723        -0.288653   
1166        1.216598        -0.175535        -0.973018        -0.288653   
216         1.084115         0.899214         1.101808        -0.288653   
970        -0.737526        -0.574410        -0.973018        -0.288653   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__MiscFeature_Not_avail  \
880        -1.035147          0.074268  ...                         1.0   
605        -0.573311         -0.659961  ...                         1.0   
1166        2.550871          1.451518  ...                         1.0   
216        -0.174865          0.863222  ...                         1.0   
970         0.345832         -0.769412  ...                         1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  \
880                     0.0                    0.0                    0.0   
605                     0.0                    0.0                    0.0   
1166                    0.0                    0.0                    0.0   
216                     0.0                    0.0                    0.0   
970                     0.0                    0.0                    0.0   

      cat__SaleCondition_Abnorml  cat__SaleCondition_AdjLand  \
880                          0.0                         0.0   
605                          0.0                         0.0   
1166                         0.0                         0.0   
216                          0.0                         0.0   
970                          1.0                         0.0   

      cat__SaleCondition_Alloca  cat__SaleCondition_Family  \
880                         0.0                        0.0   
605                         0.0                        0.0   
1166                        0.0                        0.0   
216                         0.0                        0.0   
970                         0.0                        0.0   

      cat__SaleCondition_Normal  cat__SaleCondition_Partial  
880                         1.0                         0.0  
605                         1.0                         0.0  
1166                        1.0                         0.0  
216                         1.0                         0.0  
970                         0.0                         0.0  

[5 rows x 154 columns]

In [38]:
ytrain.head()

SalePrice
710       52000
1098     128000
1286     143000
992      187000
631      209500

In [39]:
ytest.head()

SalePrice
880      157000
605      205000
1166     245350
216      210000
970      135000

### Step 6: build and evaluate the model

#### 1. linear

In [40]:
from sklearn.linear_model import LinearRegression
model1 = LinearRegression()
model1.fit(xtrain, ytrain)

LinearRegression()

In [41]:
model1.score(xtrain, ytrain)

0.9269248461730698

In [43]:
model1.score(xtest, ytest)

0.8187874407287494

#### 2. Ridge

In [45]:
from sklearn.linear_model import Ridge
model2 = Ridge(alpha=2)
model2.fit(xtrain, ytrain)

Ridge(alpha=2)

In [46]:
model2.score(xtrain, ytrain)

0.9055884144412811

In [47]:
model2.score(xtest, ytest)

0.8402027301678193

In [48]:
import numpy as np

alphas = np.arange(start=0.1, stop=100, step=0.1)
alphas

array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
       11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
       12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
       13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
       14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15. , 15

In [49]:
params = {'alpha' : alphas}
params

{'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
         5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
         6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
         7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
         8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
        10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
        11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
        12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
        13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
        14.4, 14.5, 14.6, 14.

In [50]:
from sklearn.model_selection import GridSearchCV
ridge_model = Ridge()
gscv_ridge = GridSearchCV(estimator=ridge_model, param_grid=params, cv=5, scoring='r2')
gscv_ridge.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7....
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [51]:
gscv_ridge.best_params_

{'alpha': np.float64(19.700000000000003)}

In [52]:
gscv_ridge.best_score_

np.float64(0.8337275634115983)

In [53]:
best_ridge = gscv_ridge.best_estimator_
best_ridge

Ridge(alpha=np.float64(19.700000000000003))

In [54]:
best_ridge.score(xtrain, ytrain)

0.8858178366233045

In [55]:
best_ridge.score(xtest, ytest)

0.835551906498985

#### 3.Lasso

In [56]:
from sklearn.linear_model import Lasso
model3 = Lasso(alpha=1)
model3.fit(xtrain, ytrain)

Lasso(alpha=1)

In [57]:
model3.score(xtrain, ytrain)

0.9269234344978414

In [58]:
model3.score(xtest, ytest)

0.8193724836498222

In [59]:
model_lasso = Lasso()
gscv_lasso = GridSearchCV(estimator=model_lasso, param_grid=params, cv=5, scoring='r2')
gscv_lasso.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7....
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [60]:
gscv_lasso.best_params_

{'alpha': np.float64(99.9)}

In [61]:
gscv_lasso.best_score_

np.float64(0.8472646869698716)

In [62]:
best_lasso = gscv_lasso.best_estimator_
best_lasso

Lasso(alpha=np.float64(99.9))

In [63]:
best_lasso.score(xtrain, ytrain)

0.9176640153910357

In [64]:
best_lasso.score(xtest, ytest)

0.8281357164451045

In [70]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error, r2_score

In [71]:
def evaluate_model(model, x, y):
    ypred = model.predict(x)

    mse = mean_squared_error(y, ypred)
    rmse = mse**(1/2)
    mae = mean_absolute_error(y, ypred)
    mape = mean_absolute_percentage_error(y, ypred)
    r2 = r2_score(y, ypred)

    # print all the results

    print(f'Mean Squared Error : {mse:.2f}')
    print(f'Root Mean Squared Error : {rmse:.2f}')
    print(f'Mean Absolute Error : {mae:.2f}')
    print(f'Mean Absolute Percentage Error : {mape:.4f}')
    print(f'R2 score : {r2:.4f}')

In [66]:
model1

LinearRegression()

In [67]:
best_ridge

Ridge(alpha=np.float64(19.700000000000003))

In [68]:
best_lasso

Lasso(alpha=np.float64(99.9))

In [72]:
for i in [model1, best_ridge, best_lasso]:
    print(type(i))
    print(f'Training Results')
    evaluate_model(i, xtrain, ytrain)
    print(f'Testing Results')
    evaluate_model(i, xtest, ytest)
    print('\n=================================================\n')

<class 'sklearn.linear_model._base.LinearRegression'>
Training Results
Mean Squared Error : 452256684.50
Root Mean Squared Error : 21266.33
Mean Absolute Error : 14547.97
Mean Absolute Percentage Error : 0.0873
R2 score : 0.9269
Testing Results
Mean Squared Error : 1228134049.43
Root Mean Squared Error : 35044.74
Mean Absolute Error : 16633.40
Mean Absolute Percentage Error : 0.0965
R2 score : 0.8188


<class 'sklearn.linear_model._ridge.Ridge'>
Training Results
Mean Squared Error : 706664905.01
Root Mean Squared Error : 26583.17
Mean Absolute Error : 15997.25
Mean Absolute Percentage Error : 0.0946
R2 score : 0.8858
Testing Results
Mean Squared Error : 1114516034.68
Root Mean Squared Error : 33384.37
Mean Absolute Error : 17413.76
Mean Absolute Percentage Error : 0.0992
R2 score : 0.8356


<class 'sklearn.linear_model._coordinate_descent.Lasso'>
Training Results
Mean Squared Error : 509571276.47
Root Mean Squared Error : 22573.69
Mean Absolute Error : 14999.70
Mean Absolute Percentage

#### Here I am going to choose Ridge model for out of sample prediction because it has highest r2 score in test 0.8356

### out of sample predictions 

In [73]:
xnew = pd.read_csv('sample_set.csv')
xnew.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

In [74]:
pre2

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('impute', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['MSSubClass', 'LotArea', 'OverallQual',
                                  'OverallCond', 'YearBuilt', 'MasVnrArea',
                                  'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF',
                                  'TotalBsmtSF', 'GrLivArea', 'BsmtFullBath',
                                  'TotRmsAbvGrd', 'Fireplaces', 'GarageCars',
                                  'WoodDeckSF', 'EnclosedPorch', 'ScreenPorch...
                                                                strategy='constant')),
                                                 ('ohe',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 ['LandContour', 'Neighborhood', 'BldgType',
                                  'HouseStyle', 'RoofMatl', 'Exterior1st',
                                  'MasVnrType', 'ExterQual', 'BsmtQual',
                                  'BsmtCond', 'BsmtExposure', 'HeatingQC',
                                  'KitchenQual', 'Functional', 'FireplaceQu',
                                  'GarageFinish', 'GarageQual', 'PavedDrive',
                                  'MiscFeature', 'SaleCondition'])])

In [75]:
xnew_pre = pre2.transform(xnew)
xnew_pre


num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
0           -0.872563      0.110763         -0.795151          0.381743   
1           -0.872563      0.375850         -0.071836          0.381743   
2            0.073375      0.332053         -0.795151         -0.517200   
3            0.073375     -0.054002         -0.071836          0.381743   
4            1.492282     -0.552407          1.374795         -0.517200   
...               ...           ...               ...               ...   
1454         2.438219     -0.859988         -1.518467          1.280685   
1455         2.438219     -0.864197         -1.518467         -0.517200   
1456        -0.872563      0.950423         -0.795151          1.280685   
1457         0.664586     -0.007600         -0.795151         -0.517200   
1458         0.073375     -0.089180          0.651479         -0.517200   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
0          -0.340077        -0.574410         0.053428         0.604293   
1          -0.439440         0.023903         1.051363        -0.288653   
2           0.852269        -0.574410         0.761852        -0.288653   
3           0.885390        -0.463612         0.347326        -0.288653   
4           0.686666        -0.574410        -0.396190        -0.288653   
...              ...              ...              ...              ...   
1454       -0.041991        -0.574410        -0.973018        -0.288653   
1455       -0.041991        -0.574410        -0.420316        -0.288653   
1456       -0.373198        -0.574410         1.711535        -0.288653   
1457        0.686666        -0.574410        -0.233889        -0.288653   
1458        0.719786        -0.053656         0.689474        -0.288653   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__MiscFeature_Not_avail  \
0          -0.672923         -0.400017  ...                         1.0   
1          -0.365032          0.619239  ...                         0.0   
2          -0.974021         -0.295127  ...                         1.0   
3          -0.550672         -0.299687  ...                         1.0   
4           1.018211          0.507509  ...                         1.0   
...              ...               ...  ...                         ...   
1454       -0.048086         -1.166169  ...                         1.0   
1455       -0.618589         -1.166169  ...                         1.0   
1456       -1.284176          0.379817  ...                         1.0   
1457        0.017567         -0.331610  ...                         0.0   
1458       -0.745367         -0.140072  ...                         1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  \
0                       0.0                    0.0                    0.0   
1                       0.0                    0.0                    0.0   
2                       0.0                    0.0                    0.0   
3                       0.0                    0.0                    0.0   
4                       0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
1454                    0.0                    0.0                    0.0   
1455                    0.0                    0.0                    0.0   
1456                    0.0                    0.0                    0.0   
1457                    0.0                    1.0                    0.0   
1458                    0.0                    0.0                    0.0   

      cat__SaleCondition_Abnorml  cat__SaleCondition_AdjLand  \
0                            0.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0

In [76]:
preds = best_ridge.predict(xnew_pre)
preds

array([[119559.54071404],
       [147800.43706085],
       [169161.81996341],
       ...,
       [176403.61319131],
       [106316.54503473],
       [226432.24300013]])

In [77]:
df_final = xnew[['Id']]
df_final

Id
0     1461
1     1462
2     1463
3     1464
4     1465
...    ...
1454  2915
1455  2916
1456  2917
1457  2918
1458  2919

[1459 rows x 1 columns]

In [78]:
df_final['SalePrice_pred'] = preds
df_final

Id  SalePrice_pred
0     1461   119559.540714
1     1462   147800.437061
2     1463   169161.819963
3     1464   182140.982646
4     1465   194200.941405
...    ...             ...
1454  2915    81295.549181
1455  2916    75218.611447
1456  2917   176403.613191
1457  2918   106316.545035
1458  2919   226432.243000

[1459 rows x 2 columns]

In [79]:
df_final.to_csv('Results.csv', index=False)